### PR TITLE
ENH Improve BOFT forward pass, remove CUDA kernel

### PR DESCRIPTION
**Pull request by BenjaminBossan**
## Description

This PR improves the forward pass of BOFT, which right now is very slow and relatively memory hungry.

BOFT provides its own CUDA kernel, but that wasn't working for quite some time. It is in fact possible to fix it, but by replacing `input.type()` with `input.scalar_type()` but even BOFT is still very slow. In the PEFT benchmarks, BOFT runtime is ~5-6 times higher than LoRA/OFT.

This PR provides an implementation that doesn't materialize the updated weights but works on the smaller  activations, LoRA-style. The change was implemented by Claude.

## Backwards compatibility

Backwards compatibility

To check backwards compatibility, I ran:

`pytest tests/regression/test_state_dict.py --regression -k boft`

which passes. Moreover, BC was checked with this script:

<details>

<summary>Comparison script</summary>

```python
"""Compare BOFT model outputs before and after a code change.

Example:

    PYTHONPATH=src python scripts/compare-boft-outputs.py --device cuda:1
"""

import argparse
import sys
import tempfile
from pathlib import Path

import torch
from transformers import AutoModelForCausalLM, AutoModelForImageClassification

from peft import BOFTConfig, get_peft_model


STORE_PATH_TEMPLATE = str(Path(tempfile.gettempdir()) / "boft-output-check-{dtype}.pt")
MODEL_ID = "facebook/opt-125m"
CONV_MODEL_ID = "peft-internal-testing/tiny-random-ResNetForImageClassification"
SEED = 0
# absolute tolerances on the logits, scaled to the precision of the dtype
DTYPE_ATOL = {"float64": 1e-10, "float32": 1e-4, "float16": 1e-2, "bfloat16": 1e-1}


def get_cases():
    # init_weights=False, otherwise the rotations are the identity and the comparison would be trivial.
    # The targeted convolutions of the ResNet are 1x1 convs whose flattened kernel dimensions are 64/10/20/30, so
    # boft_block_size=2 is the only non-trivial choice that fits all of them; the butterfly case targets only the
    # stage 0 conv (dimension 64), as the other dimensions cannot be split into an even number of even-sized blocks.
    return {
        "linear-defaults":    ("opt", BOFTConfig(init_weights=False)),
        "linear-block-num-8": ("opt", BOFTConfig(boft_block_size=0, boft_block_num=8, init_weights=False)),
        "linear-butterfly-2": ("opt", BOFTConfig(boft_block_size=8, boft_n_butterfly_factor=2, init_weights=False)),
        "conv-defaults":      ("resnet", BOFTConfig(target_modules=["0.layer.0.convolution"], boft_block_size=2, init_weights=False)),
        "conv-butterfly-2":   ("resnet", BOFTConfig(target_modules=["stages.0.layers.0.layer.0.convolution"], boft_block_size=4, boft_n_butterfly_factor=2, init_weights=False)),
    }


def compute_outputs(kind, config, device, dtype):
    torch.manual_seed(SEED)
    if kind == "opt":
        model = AutoModelForCausalLM.from_pretrained(MODEL_ID)
        inputs = torch.arange(20).view(2, 10)
    else:
        model = AutoModelForImageClassification.from_pretrained(CONV_MODEL_ID)
        inputs = torch.randn(2, 3, 32, 32).to(dtype)

    model = model.to(dtype)
    # autocast_adapter_dtype=False so that for float16/bfloat16, the adapter itself runs in that dtype
    model = get_peft_model(model, config, autocast_adapter_dtype=False).to(device).eval()
    inputs = inputs.to(device)

    def forward():
        with torch.inference_mode():
            output = model(inputs)
        return (output.logits if hasattr(output, "logits") else output).double().cpu()

    unmerged = forward()
    model.merge_adapter()
    merged = forward()

    # Record the boft_R shape: old versions of the code silently forced boft_n_butterfly_factor to 1 when the CUDA
    # extension could not be loaded, in which case the butterfly cases are not comparable across code versions.
    boft_R = next(param for name, param in model.named_parameters() if "boft_R" in name)
    return {"unmerged": unmerged, "merged": merged, "boft_R_shape": tuple(boft_R.shape)}


def compare_case(name, stored, current, atol):
    """Compare stored and current outputs of one case, return False on mismatch."""
    if stored["boft_R_shape"] != current["boft_R_shape"]:
        print(
            f"{name}: SKIPPED -- boft_R shape changed from {stored['boft_R_shape']} to {current['boft_R_shape']}; "
            "the reference was probably created with boft_n_butterfly_factor forced to 1 because the CUDA extension "
            "could not be loaded, so the outputs are not comparable"
        )
        return True

    ok = True
    for key in ("unmerged", "merged"):
        diff = (stored[key] - current[key]).abs()
        max_abs_diff = diff.max().item()
        max_rel_diff = (diff / (stored[key].abs() + 1e-8)).max().item()
        if max_abs_diff > atol:
            ok = False
        status = "OK" if max_abs_diff <= atol else "MISMATCH"
        print(f"{name} ({key}): {status} -- max abs diff {max_abs_diff:.3e}, max rel diff {max_rel_diff:.3e}")
    return ok


def run_dtype(dtype_name, device, reset):
    """Store or compare the outputs of all cases for one dtype, return False on mismatch."""
    store_path = Path(STORE_PATH_TEMPLATE.format(dtype=dtype_name))
    dtype = getattr(torch, dtype_name)
    results = {name: compute_outputs(kind, config, device, dtype) for name, (kind, config) in get_cases().items()}

    if reset or not store_path.exists():
        torch.save(results, store_path)
        print(f"[{dtype_name}] Stored reference outputs for {len(results)} cases at {store_path}")
        return True

    stored = torch.load(store_path, weights_only=True)
    atol = DTYPE_ATOL[dtype_name]
    print(f"[{dtype_name}] Comparing against reference outputs from {store_path} (atol {atol:.0e})")
    all_ok = True
    for name, current in results.items():
        if name not in stored:
            print(f"{name}: no stored reference outputs found")
            continue
        all_ok &= compare_case(name, stored[name], current, atol)
    print()
    return all_ok


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--device", default="cuda" if torch.cuda.is_available() else "cpu")
    parser.add_argument(
        "--dtypes",
        nargs="+",
        default=["float32", "float16", "bfloat16"],
        choices=list(DTYPE_ATOL),
        help="dtypes of the model to check; float64 distinguishes real differences from fp reordering noise",
    )
    parser.add_argument("--reset", action="store_true", help="store fresh reference outputs")
    args = parser.parse_args()

    all_ok = True
    for dtype_name in args.dtypes:
        all_ok &= run_dtype(dtype_name, args.device, args.reset)

    if all_ok:
        print("All comparable outputs match.")
    else:
        print("Some outputs DO NOT match, see above.")
        sys.exit(1)


if __name__ == "__main__":
    main()
```

</details>

It generally passes, with one caveat:

```
linear-block-num-8 (unmerged): MISMATCH -- max abs diff 1.875e-01, max rel diff 1.023e+04
```

This is, however, a floating point accumulation issue, the same passes for float32 and float16, as well as for the merged adapter.

Checking the MetaMathQA benchmark, on my machine, runtime decreased considerably:

- before:

```json
      {
        "step": 250,
        "valid accuracy": 0.0,
        "train loss": 1.2913684501647948,
        "train samples": 1000,
        "loss avg": 1.2913684501647948,
        "train time": 106.86943660605903,
        "eval time": 72.25067273300374,
        "tokens / sec": 1981.099617661842,
        "mem allocated avg": 6794620831.744,
        "mem reserved avg": 14870384738.304,
        "elapsed time": 237.60054448700248
      }
```

- after:

```json
      {
        "step": 250,
        "valid accuracy": 0.0,
        "train loss": 1.2911403470039369,
        "train samples": 1000,
        "train time": 46.50504248816287,
        "eval time": 12.977993234002497,
        "tokens / sec": 4552.6030871574785,
        "mem allocated avg": 6791611840.512,
        "mem reserved avg": 14236340191.232,
        "elapsed time": 80.2869570540206
      }
```

Comparing the train loss of a full run (from the GPU runner vs local
on the branch):

| train loss before | train loss after | diff      |
|-------------------|------------------|-----------|
| 1.291140          | 1.291222         | -0.000081 |
| 1.065423          | 1.065664         | -0.000241 |
| 0.876090          | 0.876172         | -0.000081 |
| 0.818606          | 0.818703         | -0.000097 |
| 0.796836          | 0.797038         | -0.000202 |
| 0.776865          | 0.776951         | -0.000086 |
| 0.763969          | 0.763982         | -0.000013 |
| 0.757478          | 0.757527         | -0.000049 |
| 0.748042          | 0.748076         | -0.000034 |
| 0.745327          | 0.745239         | 0.000088  |
| 0.736915          | 0.736921         | -0.000006 |
| 0.728413          | 0.728494         | -0.000080 |
| 0.736114          | 0.736053         | 0.000061  |
| 0.724543          | 0.724551         | -0.000008 |
| 0.719788          | 0.719767         | 0.000020  |
| 0.738617          | 0.738715         | -0.000098 |
| 0.716734          | 0.716832         | -0.000098 |
| 0.727882          | 0.727911         | -0.000028 |
| 0.720758          | 0.720830         | -0.000071 |
| 0.726906          | 0.726833         | 0.000073  |
| 0.726906          | 0.726833         | 0.000073  |

Thus it looks pretty much like the losses are identical, whereas runtime is greatly improved and memory slightly improved.

For image-gen, I got an OOM error with current main. On the new branch, training started successfully. I'm thus comparing a local run on this branch to the results from the GPU runner:

- before:

```json
      {
        "step": 100,
        "valid dino_similarity": 0.34351247549057007,
        "train loss": 0.762182354927063,
        "train samples": 200,
        "train time": 243.34609632800084,
        "eval time": 66.00437694799984,
        "mem allocated avg": 14321960192.0,
        "mem reserved avg": 17365215477.76,
        "elapsed time": 315.7438521270001
      }
```

- after:

```json
      {
        "step": 100,
        "valid dino_similarity": 0.5274292230606079,
        "train loss": 0.7614542841911316,
        "train samples": 200,
        "train time": 113.77418193017365,
        "eval time": 13.306961588998092,
        "mem allocated avg": 8310736128.0,
        "mem reserved avg": 10452205568.0,
        "elapsed time": 132.47501685799216
      }
```

So on the image-gen benchmark, it seems we get similar improvements in terms of runtime and memory.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
@githubnemo I applied the skill from #3544 to this PR.

**Comment by BenjaminBossan** [comment]
@yfeng95 If you have the opportunity, it would be great if you could check this PR.

**Comment by BenjaminBossan** [comment]
_not stale_
